homeassistant/components/fyta/sensor.py

```diff
@@ -0,0 +1,168 @@
+
+SENSORS: Final[list[FytaSensorEntityDescription]] = [
+    FytaSensorEntityDescription(
+        key="scientific_name",
+        translation_key="scientific_name",
+    ),
+    FytaSensorEntityDescription(
+        key="status",
+        translation_key="plant_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=PLANT_STATUS_LIST,
+        value_fn=lambda value: PLANT_STATUS[value],
+    ),
+    FytaSensorEntityDescription(
+        key="temperature_status",
+        translation_key="temperature_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=PLANT_STATUS_LIST,
+        value_fn=lambda value: PLANT_STATUS[value],
+    ),
+    FytaSensorEntityDescription(
+        key="light_status",
+        translation_key="light_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=PLANT_STATUS_LIST,
+        value_fn=lambda value: PLANT_STATUS[value],
+    ),
+    FytaSensorEntityDescription(
+        key="moisture_status",
+        translation_key="moisture_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=PLANT_STATUS_LIST,
+        value_fn=lambda value: PLANT_STATUS[value],
+    ),
+    FytaSensorEntityDescription(
+        key="salinity_status",
+        translation_key="salinity_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=PLANT_STATUS_LIST,
+        value_fn=lambda value: PLANT_STATUS[value],
+    ),
+    FytaSensorEntityDescription(
+        key="temperature",
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        device_class=SensorDeviceClass.TEMPERATURE,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="light",
+        translation_key="light",
+        native_unit_of_measurement="mol/d",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="moisture",
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=SensorDeviceClass.MOISTURE,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="salinity",
+        translation_key="salinity",
+        native_unit_of_measurement="mS/cm",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="ph",
+        device_class=SensorDeviceClass.PH,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="battery_level",
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=SensorDeviceClass.BATTERY,
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+]
+
+PLANT_NUMBER_SENSOR = FytaSensorEntityDescription(
+    key="plant_number",
+    translation_key="plant_number",
+    state_class=SensorStateClass.MEASUREMENT,
+    entity_category=EntityCategory.DIAGNOSTIC,
+)
```

Comment:
Okay I just had a conversation with someone else about this sensor, and currently it's a calculated state, which we don't really allow. My proposal is to remove it in this PR, and have it added in a separate PR where you can express why this sensor should be included and what the use case is.

Reply:
Wait. Should I now also delete this sensor? I begin wondering, if you are interested in the integration at all. 

I know that you as well are doing this in your spare time. But this also applies for me. I'm not able to rewrite and rewrite ad libidum. If I have to reduce code until there is almost nothing remaining, I'm wondering if I should not better invest in further developing just the custom component that at least I can use.

Reply:
> I definetly am interested in the integration. I don't use it, but I like to add value to the platform we love.
> 
> I have a hard time trying to figure out what your plan was with the controller entity from the beginning, maybe it got burried under other comments, but I tried to uncover what entities you were planning to add and from your explanation I got the idea that the controller also could have an illuminance, temperature and what not.
> 
> I mean we are almost there, what do you think is the best way forward. We could list the controller device entities you wanted to add and I can get someone to say if they should or could be included. If not, we should refactor the code to only add the parent device as device with no entities.
> 
> You always have the option to go for a custom integration, but after a week of back and forth I think it would be sad not to include it in core.

It really would be a pitty after all the work done, if the integration would not find its way into core. However, work performed is not a factor to decide the next steps (sunken costs must not be considered in investment decisions). And frankly, I don't think I'm willing to invest anymore time in removing functionality (which entails adjusting code here and there and testing allover again). Already now the functionality is below my expectation. And thinking that anything now removed has to be redone in a seperate PR with overdoing the whole procedure again and having to justify why I think a feature is helpful (anyone not needing it can disable it!)... Not sure if this is what I want to do with my free time. 

I fully understand that code integrity must be saveguarded and also appreciate any hint on how the code can be improved, in particular as I'm new to Python. Unfortunately the (undocumented!) HA-standards apparently go way beyond this.

I suggest you decide if you can live with the current state of functionality and I'm more than happy to push it over the line. If not, I will look, if I find the motivation to pursue this PR or if I prefer spending my energy more efficiently.


PS It would save you and any possible first contributor a lot of work, if not only the basic functionalities but all hidden requirenents (we don't like this and that sensor, don't include this in the PR...) would be documented and good examples for specific use cases would be mentioned. I took quite a lot of inspirations from integrations that apparently do not suffice the current standards anymore, which also is quite anoying.

Reply:
> Okay I just had a conversation with someone else about this sensor, and currently it's a calculated state, which we don't really allow. My proposal is to remove it in this PR, and have it added in a separate PR where you can express why this sensor should be included and what the use case is.

The number of plants is not calculated, but counted...

Reply:
But can you still respond to the question I asked? What entities would you want to see in the controller device. With this information i will ask around to see if I'm just doing a bad job here or that we should indeed change it.

Like this is now making me doubt the way on how I approach these reviews and I want to verify my findings with others so I can improve.

And if you don't want to invest more time, I'm willing to help you with the changes to drag it across the finish line. I think we had a fruitful week with making the integration up to standards.

And also please remember, we like functionality, but it should make sense. So if something could be added that's already covered by another way in core, we should not do it, since it's already covered.

And I agree that the document is not complete. I still play around with the idea of experimenting with documentation and writing down what I have learnt in the last 10 months. But since I'm currently in the middle of my graduation, my time for these things is limited.

Reply:
From the top of my head, the following entities might come up:
- Timestamp, when latest data was received (data is only pushed every 30? or so min. to server, so it would be helpful to see how old the data is)
- Hub (the Beam sensors connect with BT to Hub, that is in the WiFi. Habe to figure out if it make sense to have a separate device for Hub or integrate in coordinator)
- State overall and for temp/light etc. (overall status of plants; would be calculated somehow, so I guess only permissible in custom integration)
- Information on gardens (those can be used to group plants)
- possible new features in FYTA API

Reply:
@joostlek, is there any news on the above? May I set the conversation as resolved?

Reply:
> State overall and for temp/light etc. (overall status of plants; would be calculated somehow, so I guess only permissible in custom integration)

This is already possible with the `group` integration directly in HA and should not be done by integrations as we don't know how the user wants to group them. There are endless possibilities and the user can do it for his needs.

> Timestamp, when latest data was received (data is only pushed every 30? or so min. to server, so it would be helpful to see how old the data is)

You can find use HA directly to find the datetime, when the state of an entity has changed. No need to provide an additional data point. Not sure if the api supports it, but it is possible to push the data async to HA, so we don't need to poll it on a fixed interval.

We can always add sensor in the future (even refactor the code), but we should not add code for thing, which will be added in the future.

I understand that you are using your spare time to develop this integration but we need also to make sure that it has certain requirements/standards as the HA community is expecting this.

I'm looking at your PR for the first time, so I can't say what was already removed from previous reviews, but if we ask to remove or to move code into a follow-up, we do that to ensure the PRs are small and have all the same quality across the whole code base.

You are currently already creating one device per plant, so in my opinion it is useless to count them and expose them as sensor as the user can retrieve this information from the device overview. Or do I miss something?

Reply:
> > 
> I'm looking at your PR for the first time, so I can't say what was already removed from previous reviews, but if we ask to remove or to move code into a follow-up, we do that to ensure the PRs are small and have all the same quality across the whole code base.
> 
> You are currently already creating one device per plant, so in my opinion it is useless to count them and expose them as sensor as the user can retrieve this information from the device overview. Or do I miss something?

As set out above, devices are only created by plants with sensors (FYTA Beam). However, the user can also add plants without sensor, that are also counted here. Therefore the number in my view is not redundant.

Reply:
> You can find use HA directly to find the datetime, when the state of an entity has changed. No need to provide an additional data point. Not sure if the api supports it, but it is possible to push the data async to HA, so we don't need to poll it on a fixed interval.

The time a state has changed is not necessarily the same as the time the sensor has pushed its values to the server for the last time. If I have a stable environment, the values of the sensors might not change for a day or more. Nevertheless it (at least for me) is helpful to see, how recent the data of the server is (for example it could give an indication if my Beam sensor became offline and could not upload its data). However, as this sensor already has been deleted in the current PR, this is a pure hypothetical discussion.

Reply:
> As set out above, devices are only created by plants with sensors (FYTA Beam). However, the user can also add plants without sensor, that are also counted here. Therefore the number in my view is not redundant.

Looking at the code we iterate over `coordinator.fyta.plant_list`, which will be used for the counting sensor too. 
I assume each sensor has at least a battery level or status so a device would be created for it.

> However, the user can also add plants without sensor, that are also counted here

How can a user add them and what's the benefit of adding them (without having the other sensor entities)? And what would be a real use case to use the plant number in HA?

Counting is calculating and we don't accept calculating entities from integrations, that are communicating with devices/services (if the calculating/counting is done directly in HA). We can make an exception if there is a good use case for the majority of the users

Reply:
> Looking at the code we iterate over coordinator.fyta.plant_list, which will be used for the counting sensor too.
As already explained several times, in plant_list there are all plants, also those without sensors.

> I assume each sensor has at least a battery level or status so a device would be created for it.
Your assumption and/or wording is wrong. If plant has no BEAM to upload data, there is also no sensor for battery_level or status.

> How can a user add them and what's the benefit of adding them (without having the other sensor entities)? 
Users can administer plants in the FYTA app. There is quite a lot functionality also for plants without sensor. I do have several plants in my App without sensors. With regards to the benefit you should ask FYTA and not me, as they provide this functionality.

Anyway. I'm tired of trying to explain over and over, why I think this sensor is helpful. I removed the respective sensor to make you all happy.